### ISSUE TITLE
Placing Conditional Orders

### ISSUE THREAD

**Issue by jeanstkng** [stoqey/ib#66]
Hi, 

I'm getting errors when I try to place an Order using the conditions property. I've tried using Time Condition, Price Condition and Margin Condition but it just doesn't works.

I've been reading and logging the encoder code but I don't see nothing wrong there. Maybe it is a problem from TWS? Am I placing the order in a wrong way?

Below is the code that throws error

```
import IBApi, { ConjunctionConnection, Contract, ErrorCode, EventName, ExecutionCondition, MarginCondition, Order, OrderAction, OrderState, OrderType, SecType, TimeCondition } from '../ib-master/dist/index';

const ib = new IBApi({
    // clientId: 0,
    // host: '127.0.0.1',
    port: 7497,
});

function handleOrderPlacement() {

    ib
    .on(EventName.error, (error: Error, code: ErrorCode, reqId: number) => {
        console.log(`Error ${error.name} msg: ${error.message} - Code ${code} - reqId ${reqId}`, error.stack)
    })
    .once(EventName.nextValidId, (orderId: number) => {
        const contract: Contract = {
            symbol: "A",
            exchange: "NYSE",
            currency: "USD",
            secType: SecType.STK
        };

        const timeCondition: TimeCondition = new TimeCondition("20210118 23:59:59 EST", true, ConjunctionConnection.OR);
        // const marginCondition: MarginCondition = new MarginCondition(10, true, ConjunctionConnection.OR);
        // const execCondition: ExecutionCondition = new ExecutionCondition("ISLAND", SecType.STK, "FB", ConjunctionConnection.OR);
        const order: Order = {
            orderType: OrderType.LMT,
            action: OrderAction.BUY,
            lmtPrice: 1,
            orderId,
            totalQuantity: 1,
            account: 'DU3746845',
            conditionsIgnoreRth: true,
            conditionsCancelOrder: false
        };
        order.conditions = [timeCondition];
        console.log(order);
        ib.placeOrder(orderId, contract, order);
    });
}

handleOrderPlacement();
ib.connect();
ib.reqIds();
```

The errors are these:

```
Error Error msg: Error reading request.Message id 135.  Unable to parse data. java.lang.NumberFormatException: For input string: "o" - Code 320 - reqId 135 Error: Error reading request.Message id 135.  Unable to parse data. java.lang.NumberFormatException: For input string: "o"

Error Error msg: Error validating request.-'bX' : cause - Please enter exchange - Code 321 - reqId 0 Error: Error validating request.-'bX' : cause - Please enter exchange
```

Thanks in advance

**Comment by mfrener** [comment]
Thanks for reporting!
Think is actually is a problem in our side. Checked the code again and we did not send the condition type, which makes sense according to error message, because what follows is the conjunctionConnection, which is 'o' and so TWS throws error because it wants an integer the (condition type)

Create a PR for it https://github.com/stoqey/ib/pull/67
If you to try it w/o waiting for review /merge/lib release: dist/core/io/encoder.js - line 953
```   
(_v = order.conditions) === null || _v === void 0 ? void 0 : _v.forEach((cond) => {
                    tokens.push(cond.conjunctionConnection);
```  
change to 
If you to try it w/o waiting for review /merge/lib release: dist/core/io/encoder.js - line 953
```   
(_v = order.conditions) === null || _v === void 0 ? void 0 : _v.forEach((cond) => {
                    tokens.push(cond.type);
                    tokens.push(cond.conjunctionConnection);
```

**Comment by mfrener** [comment]
Release is already available, please you version >

I have re-opened it to wait for confirmation before finally closed.
I can see on Gateway logs that an order has been received and there is a condition on it, but have not checked other types and also not verified if it did actually add correctly to the order-list and is executed properly 

From gateway:
09:34:15:618 <- 3-1--A-STK-----NYSE--USD-----BUY-1-LMT-1----DU3387025-----------------------------------------------------------0--------0--------------------
09:55:03:970 <- 3-2--A-STK-----NYSE--USD-----BUY-1-LMT-1----DU3387025-----------------------------------------------------------0--------0--------------------
09:55:54:608 <- 3-3--A-STK-----NYSE--USD-----BUY-1-LMT-1----DU3387025-----------------------------------------------------------0--------0--------------------
09:56:26:710 <- 3-4--A-STK-----NYSE--USD-----BUY-1-LMT-1----DU3387025-----------------------------------------------------------0--------1-3-o-0-20210118 23:59:59 EST-1-0--------------------
09:56:50:528 <- 3-5--A-STK-----NYSE--USD-----BUY-1-LMT-1----DU3387025-----------------------------------------------------------0--------1-3-o-0-20210118 23:59:59 EST-1-0--------------------

Keeping it open until confirmation from @jeanstkng  that works now

**Comment by jeanstkng** [comment]
I also thought that could be the problem but I didn't knew where to push it.
It works now for all conditions, thank you @mfrener!

**Comment by jeanstkng** [comment]
I got this error, now the problem comes from decoding.
The order is transmitted but this error is logged.

```
Decoding error on OPEN_ORDER: unprocessed data left on queue (["0","1","0","0","1",null]). Please report to https://github.com/stoqey/ib - code: 505 - reqId: -1 Error: Decoding error on OPEN_ORDER:
unprocessed data left on queue (["0","1","0","0","1",null]). Please report to https://github.com/stoqey/ib
```

**Comment by jeanstkng** [comment]
It might be that the condition type is not being decoded so its being left on queue

**Comment by mfrener** [comment]
Found it, PP is open https://github.com/stoqey/ib/pull/70
Was a copy&past error from java porting.... conditionsIgnoreRth and conditionsCancelOrder need to be read after all conditions are read, not inside the loop after each condition.
Tested on real TWS with your code from above, but using all 3 conditions and a reqOpenOrders() to pull it again.

**Comment by mfrener** [comment]
Fix is available with 1.1.31.
btw. checkout out IBApiNext if you want to get rid of that event-callback hell - there is no `getOpenOrders(): Observable<OpenOrder[]>` yet, but it's next on todo list (as various others already asked for it ;))
